historydag/parsimony.py

```diff
@@ -0,0 +1,653 @@
+"""A module implementing Sankoff Algorithm"""
+
+import random
+import ete3
+import numpy as np
+import Bio.Data.IUPACData
+from historydag.utils import access_nodefield_default
+from itertools import product
+from historydag.dag import history_dag_from_clade_trees, history_dag_from_etes, HistoryDag
+
+bases = "AGCT-"
+ambiguous_dna_values = Bio.Data.IUPACData.ambiguous_dna_values.copy()
+ambiguous_dna_values.update({"?": "GATC-", "-": "-"})
+
+
+def hamming_distance(seq1: str, seq2: str) -> int:
+    r"""Hamming distance between two sequences of equal length.
+
+    Args:
+        seq1: sequence 1
+        seq2: sequence 2
+    """
+    if len(seq1) != len(seq2):
+        raise ValueError("sequence lengths do not match!")
+    return sum(x != y for x, y in zip(seq1, seq2))
+
+
+_yey = np.array(
+    [
+        [0, 1, 1, 1, 1],
+        [1, 0, 1, 1, 1],
+        [1, 1, 0, 1, 1],
+        [1, 1, 1, 0, 1],
+        [1, 1, 1, 1, 0],
+    ]
+)
+
+# This is applicable even when diagonal entries in transition rate matrix are
+# nonzero, since it is only a mask on allowable sites based on each base.
+code_vectors = {
+    code: np.array(
+        [0 if base in ambiguous_dna_values[code] else float("inf") for base in bases]
+    )
+    for code in ambiguous_dna_values
+}
+
+ambiguous_codes_from_vecs = {
+    tuple(0 if base in base_set else 1 for base in bases): code
+    for code, base_set in ambiguous_dna_values.items()
+}
+
+
+def _get_adj_array(seq_len, transition_weights=None):
+    if transition_weights is None:
+        transition_weights = _yey
+    else:
+        transition_weights = np.array(transition_weights)
+
+    if transition_weights.shape == (5, 5):
+        adj_arr = np.array([transition_weights] * seq_len)
+    elif transition_weights.shape == (seq_len, 5, 5):
+        adj_arr = transition_weights
+    else:
+        raise RuntimeError(
+            "Transition weight matrix must have shape (5, 5) or (sequence_length, 5, 5)."
+        )
+    return adj_arr
+
+
+def edge_weight_func_from_weight_matrix(dag, weight_mat, bases):
+    @access_nodefield_default("sequence", 0)
+    def wrapped_weighted_hamming_distance(s1, s2) -> int:
+        """The sitewise sum of base differences between s1 and s2."""
+        base_indices = {k: v for v, k in enumerate(bases)}
```

Comment:
should this be moved outside the wrapped function so it's built once, instead of every time the returned function is called? Not a big deal of course

Reply:
Good point! Yes. That would be more efficient.